File: ProblemSolving/450DSA/Python/src/arrays/MaximumProductSubarray.py
"""
@author Anirudh Sharma

Given an integer array nums, find the contiguous subarray within an array
(containing at least one number) which has the largest product.
"""


def maxProduct(nums):
    # Special cases
    if nums is None or len(nums) == 0:
        return -1
    # Overall maximum product
    globalMaxima = nums[0]
    # Maximum product until the current index
    localMaxima = nums[0]
    # Minimum product until the current index
    localMinima = nums[0]
    # Loop for the remaining elements
    for i in range(1, len(nums)):
        # Save localMaxima for localMinima calculation
        temp = localMaxima
        localMaxima = max(nums[i], max(localMaxima * nums[i], localMinima * nums[i]))
        localMinima = min(nums[i], min(temp * nums[i], localMinima * nums[i]))
        globalMaxima = max(localMaxima, globalMaxima)
    return globalMaxima


if __name__ == "__main__":
    print(maxProduct([2, 3, -2, 4]))
    print(maxProduct([-2, 0, -1]))
    print(maxProduct([6, -3, -10, 0, 2]))
    print(maxProduct([2, 3, 4, 5, -1, 0]))

File: ProblemSolving/450DSA/JavaScript/src/arrays/MaximumProductSubarray.js
/**
 * @author Anirudh Sharma
 *
 * Given an integer array nums, find the contiguous subarray within an array
 * (containing at least one number) which has the largest product.
 */
const maxProduct = (nums) => {
    // Special cases
    if (nums === undefined || nums.length === 0) {
        return -1;
    }
    // Overall maximum product
    let globalMaxima = nums[0];
    // Maximum product until a given index
    let localMaxima = nums[0];
    // Minimum product until a given index
    let localMinima = nums[0];
    // Loop for the remaining elements in the array
    for (let i = 1; i < nums.length; i++) {
        // Save localMaxima for localMinima calculation
        let temp = localMaxima;
        localMaxima = Math.max(nums[i], Math.max(localMaxima * nums[i], localMinima * nums[i]));
        localMinima = Math.min(nums[i], Math.min(temp * nums[i], localMinima * nums[i]));
        globalMaxima = Math.max(localMaxima, globalMaxima);
    }
    return globalMaxima;
};


console.log(maxProduct([2, 3, -2, 4]));
console.log(maxProduct([-2, 0, -1]));
console.log(maxProduct([6, -3, -10, 0, 2]));
console.log(maxProduct([2, 3, 4, 5, -1, 0]));